lib/language_pack/ruby.rb

```diff
@@ -1018,6 +1018,26 @@ def add_node_js_binary
     if Pathname(build_path).join("package.json").exist? ||
          bundler.has_gem?('execjs') ||
          bundler.has_gem?('webpacker')
+
+      version = @node_installer.version
+      old_version = @metadata.fetch("default_node_version") { version }
```

Comment:
I don't see where `default_node_version` (or `default_yarn_version`) is ever written to?

Reply:
The fetch with a block is a pattern in ruby:

```
hash = {}
one = hash.fetch("a_key") { "fallback" } 
puts one.inspect
# => "fallback"
```

In this case it's backed by a cache, which is conceptually similar to `Rails.cache#fetch` https://api.rubyonrails.org/v7.1.1/classes/ActiveSupport/Cache/Store.html#method-i-fetch

The implementation for our version of `fetch` is here https://github.com/heroku/heroku-buildpack-ruby/blob/77c1ac34e7c1e8b45bf2af9008aa47d5eb783943/lib/language_pack/metadata.rb#L47-L54

Reply:
Oh that's a really unfortunately named API.

Both the name and the appearance at the call site make it seem like the "fetch a value from this dict/hash/... and if not set fallback to value" pattern. Whereas it seems instead this API tries to simultaneously fetch a value and store a different value, under an API that name that suggests it only retries a value.